http-clients/src/test/java/com/palantir/remoting/http/OptionalAwareContractTest.java

```diff
@@ -0,0 +1,133 @@
+import com.squareup.okhttp.mockwebserver.MockResponse;
+import com.squareup.okhttp.mockwebserver.MockWebServer;
+import com.squareup.okhttp.mockwebserver.RecordedRequest;
+import javax.net.ssl.SSLSocketFactory;
+import javax.ws.rs.GET;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+public final class OptionalAwareContractTest {
+
+    @Rule
+    public final MockWebServer server = new MockWebServer();
+
+    private FakeoInterface proxy;
+
+    @Before
+    public void before() {
+        proxy = FeignClients.standard().createProxy(Optional.<SSLSocketFactory>absent(),
+                ImmutableSet.of("http://localhost:" + server.getPort()),
+                FakeoInterface.class);
+        server.enqueue(new MockResponse().setBody("\"foo\""));
+    }
+
+    @Test
+    public void testAbsentPath() throws Exception {
+        proxy.path(Optional.<String>absent());
+        RecordedRequest takeRequest = server.takeRequest();
+        // note that this differs from feign's default behavior of leaving the template
+        // in the path. The default would be "/{opt}/foo"
+        // This is because feign expands null values to their parameter name if they
+        // have no Expander and the value is null
+        assertThat(takeRequest.getPath(), is("/null/foo"));
+    }
+
+    @Test
+    public void testEmptyStringPath() throws Exception {
+        proxy.path(Optional.<String>of(""));
+        RecordedRequest takeRequest = server.takeRequest();
+        assertThat(takeRequest.getPath(), is("//foo"));
+    }
+
+    @Test
+    public void testStringPath() throws Exception {
+        proxy.path(Optional.<String>of("str"));
+        RecordedRequest takeRequest = server.takeRequest();
+        assertThat(takeRequest.getPath(), is("/str/foo"));
+    }
+
+    @Test
+    public void testAbsentQuery() throws Exception {
+        proxy.query(Optional.<String>absent());
+        RecordedRequest takeRequest = server.takeRequest();
+        assertThat(takeRequest.getRequestLine(), is("GET /foo HTTP/1.1"));
+    }
+
+    @Test
+    public void testEmptyStringQuery() throws Exception {
+        proxy.query(Optional.<String>of(""));
+        RecordedRequest takeRequest = server.takeRequest();
+        assertThat(takeRequest.getRequestLine(), is("GET /foo?opt= HTTP/1.1"));
+    }
+
+    @Test
+    public void testStringQuery() throws Exception {
+        proxy.query(Optional.<String>of("str"));
+        RecordedRequest takeRequest = server.takeRequest();
+        assertThat(takeRequest.getRequestLine(), is("GET /foo?opt=str HTTP/1.1"));
+    }
+
+    @Test
+    public void testAbsentHeader() throws Exception {
+        proxy.header(Optional.<String>absent());
+        RecordedRequest takeRequest = server.takeRequest();
+        assertThat(takeRequest.getHeader("opt"), is("{opt}"));
```

Comment:
It's how the innards of Feign work. The only way to fix is to change their code too :(